crates/fee/src/lib.rs

```diff
@@ -465,55 +476,123 @@ impl<T: Config> Pallet<T> {
-                .try_into()
-                .map_err(|_| Error::<T>::TryIntoIntError)?;
-            let amount = Amount::<T>::new(rewards, currency_id);
-            amount.transfer(&Self::fee_pool_account_id(), nominator_id)?;
-        }
-        Ok(())
+        let rewards = T::VaultStaking::withdraw_reward(&(index, vault_id.clone()), nominator_id, currency_id)?
+            .try_into()
+            .map_err(|_| Error::<T>::TryIntoIntError)?;
+        let amount = Amount::<T>::new(rewards, currency_id);
+        amount.transfer(&Self::fee_pool_account_id(), nominator_id)?;
+
+        Ok(rewards)
     }
 
     fn distribute(reward: &Amount<T>) -> Result<Amount<T>, DispatchError> {
         Ok(
-            if let Err(_) = T::VaultRewards::distribute_reward(&(), reward.currency(), reward.amount()) {
+            if let Err(_) = T::CapacityRewards::distribute_reward(&(), reward.currency(), reward.amount()) {
                 reward.clone()
             } else {
                 Amount::<T>::zero(reward.currency())
             },
         )
     }
 
-    pub fn distribute_from_reward_pool(vault_id: &DefaultVaultId<T>) -> DispatchResult {
-        for currency_id in [vault_id.wrapped_currency(), T::GetNativeCurrencyId::get()] {
-            let reward = T::VaultRewards::withdraw_reward(&(), vault_id, currency_id)?;
+    fn distribute_vault_rewards(vault_id: &DefaultVaultId<T>, currency_id: CurrencyId<T>) -> DispatchResult {
+        let collateral_id = vault_id.collateral_currency();
 
-            let full_amount = Amount::<T>::new(reward, currency_id);
+        // push rewards based on collateral capacity
+        let reward = T::CapacityRewards::withdraw_reward(&(), &collateral_id, currency_id)?;
+        T::VaultRewards::distribute_reward(&collateral_id, currency_id, reward)?;
 
-            let commission_rate = Self::get_commission_rate(vault_id);
-            let commission = full_amount.checked_fixed_point_mul(&commission_rate)?;
-            commission.transfer(&Self::fee_pool_account_id(), &vault_id.account_id)?;
+        // push rewards based on vault's contribution to that capacity
+        let reward = T::VaultRewards::withdraw_reward(&collateral_id, vault_id, currency_id)?;
 
-            let remainder = full_amount.checked_sub(&commission)?;
+        let full_amount = Amount::<T>::new(reward, currency_id);
 
-            T::VaultStaking::distribute_reward(&(None, vault_id.clone()), currency_id, remainder.amount())?;
-        }
+        let commission_rate = Self::get_commission_rate(vault_id);
+        let commission = full_amount.checked_fixed_point_mul(&commission_rate)?;
+        commission.transfer(&Self::fee_pool_account_id(), &vault_id.account_id)?;
+
+        let remainder = full_amount.checked_sub(&commission)?;
+
+        T::VaultStaking::distribute_reward(&(None, vault_id.clone()), currency_id, remainder.amount())?;
 
         Ok(())
     }
 }
+
+pub struct RewardsRouter<T>(sp_std::marker::PhantomData<T>);
```

Comment:
This is my suggestion for an "atomic" wrapper type to ensure we always distribute rewards when changing stake at a lower level.